osu.Game/Screens/Edit/Timing/WaveformComparisonDisplay.cs

```diff
@@ -0,0 +1,218 @@
+            selectedGroupEndTime = beatmap.Value.Track.Length;
+
+            var tcp = selectedGroup.Value?.ControlPoints.OfType<TimingControlPoint>().FirstOrDefault();
+
+            if (tcp == null)
+            {
+                timingPoint = new TimingControlPoint();
+                return;
+            }
+
+            timingPoint = tcp;
+            beatLength.BindTo(timingPoint.BeatLengthBindable);
+
+            selectedGroupStartTime = selectedGroup.Value?.Time ?? 0;
+
+            var nextGroup = editorBeatmap.ControlPointInfo.TimingPoints
+                                         .SkipWhile(g => g != tcp)
+                                         .Skip(1)
+                                         .FirstOrDefault();
+
+            if (nextGroup != null)
+                selectedGroupEndTime = nextGroup.Time;
+        }
+
+        protected override bool OnHover(HoverEvent e) => true;
+
+        protected override bool OnMouseMove(MouseMoveEvent e)
+        {
+            float trackLength = (float)beatmap.Value.Track.Length;
+            int totalBeatsAvailable = (int)(trackLength / timingPoint.BeatLength);
+
+            Scheduler.AddOnce(showFrom, (int)(e.MousePosition.X / DrawWidth * totalBeatsAvailable));
+
+            return base.OnMouseMove(e);
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+
+            if (!IsHovered)
+            {
+                int currentBeat = (int)Math.Floor((editorClock.CurrentTimeAccurate - selectedGroupStartTime) / timingPoint.BeatLength);
+
+                showFrom(currentBeat);
+            }
+        }
+
+        private void showFrom(int beatIndex)
+        {
+            if (lastDisplayedBeatIndex == beatIndex)
+                return;
+
+            // Chosen as a pretty usable number across all BPMs.
+            // Optimally we'd want this to scale with the BPM in question, but performing
+            // scaling of the display is both expensive in resampling, and decreases usability
+            // (as it is harder to track the waveform when making realtime adjustments).
+            const float visible_width = 300;
+
+            float trackLength = (float)beatmap.Value.Track.Length;
+            float scale = trackLength / visible_width;
+
+            // Start displaying from before the current beat
+            beatIndex -= total_waveforms / 2;
+
+            foreach (var row in InternalChildren.OfType<WaveformRow>())
+            {
+                // offset to the required beat index.
+                double time = selectedGroupStartTime + beatIndex * timingPoint.BeatLength;
+
+                float offset = (float)(time - visible_width / 2) / trackLength * scale;
+
+                row.Alpha = time < selectedGroupStartTime || time > selectedGroupEndTime ? 0.2f : 1;
+                row.WaveformOffset = -offset;
+                row.WaveformScale = new Vector2(scale, 1);
+                row.BeatIndex = beatIndex++;
```

Comment:
It's a minor nitpick but I found it kinda disorienting not knowing where the current beat is in the list. I understand that that's not the point and that you should be looking around and matching all of them up, but it kinda makes me lose track of what I'm looking at. It's compounded by the current beat not being at the top of the list but somewhere in-the-middle-down-one.

A future improvement I would propose to help with this is for this index to be a constant value of `-4 -3 -2 -1 0 1 2 3` for the list, but we'll see what others say first.

Reply:
Do agree with this. I experimented with changing the colour of the line with the current beat, but it was causing invalidation of the waveform so I left that out for now.